File: realtime/ios/webrtc/Sources/Auth/AuthProvider.swift
import Foundation

struct IceServer: Decodable, Equatable {
    var urls: [String]
    var username: String?
    var credential: String?

    enum CodingKeys: String, CodingKey {
        case urls, username, credential
    }

    init(urls: [String], username: String? = nil, credential: String? = nil) {
        self.urls = urls
        self.username = username
        self.credential = credential
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        // "urls" may be a single string or an array of strings
        if let single = try? container.decode(String.self, forKey: .urls) {
            urls = [single]
        } else {
            urls = try container.decode([String].self, forKey: .urls)
        }
        username = try container.decodeIfPresent(String.self, forKey: .username)
        credential = try container.decodeIfPresent(String.self, forKey: .credential)
    }
}

struct RealtimeCredentials {
    var authorizationHeader: String
    var apiBaseURL: URL
    var callsURL: URL?
    var preFetchedIceServers: [IceServer]?
}

enum AuthError: LocalizedError {
    case missingAPIKey
    case invalidBackendURL
    case backendRequestFailed(String)

    var errorDescription: String? {
        switch self {
        case .missingAPIKey: "Inworld API key is not set. Add it in Settings."
        case .invalidBackendURL: "Backend URL is invalid."
        case .backendRequestFailed(let detail): "Backend config request failed: \(detail)"
        }
    }
}

protocol AuthProvider {
    func credentials() async throws -> RealtimeCredentials
}

File: realtime/android/webrtc/app/src/main/java/ai/inworld/voiceagent/auth/AuthProvider.kt
package ai.inworld.voiceagent.auth

import kotlinx.serialization.KSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.builtins.serializer
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.JsonTransformingSerializer

/** "urls" may arrive as a single string or an array of strings. */
private object UrlsSerializer : JsonTransformingSerializer<List<String>>(ListSerializer(String.serializer())) {
    override fun transformDeserialize(element: JsonElement): JsonElement =
        if (element is JsonPrimitive) JsonArray(listOf(element)) else element
}

@Serializable
data class IceServer(
    @Serializable(with = UrlsSerializer::class) val urls: List<String>,
    val username: String? = null,
    val credential: String? = null,
)

data class RealtimeCredentials(
    val authorizationHeader: String,
    val apiBaseUrl: String = "https://api.inworld.ai",
    val callsUrl: String? = null,
    val preFetchedIceServers: List<IceServer>? = null,
)

sealed class AuthException(message: String) : Exception(message) {
    class MissingApiKey : AuthException("Inworld API key is not set. Add it in Settings.")
    class InvalidBackendUrl : AuthException("Backend URL is invalid.")
    class BackendRequestFailed(detail: String) : AuthException("Backend config request failed: $detail")
}

fun interface AuthProvider {
    suspend fun credentials(): RealtimeCredentials
}
